### PR TITLE
Support latest TransformerEngine

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Support latest Transformer-Engine

**Major Revision**
- Upgrade TE to latest stable v0.11
- Integrate MS-AMP with TE

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -38,7 +38,7 @@ classifiers=[
 ]
 dependencies = [
     "torch",
-    "transformer-engine@git+https://github.com/NVIDIA/TransformerEngine.git@v0.7#egg=transformer-engine",
+    "transformer-engine@git+https://github.com/NVIDIA/TransformerEngine.git@v0.11#egg=transformer-engine",
```
Comment (wkcn) [inline pyproject.toml:41]: I have a question on it. There is no version requirement like `transformer-engine==v0.11`. Will TransformerEngine be re-install when installing MS-AMP even if the version of local TransformerEngine is v0.11?

File: tests/te/test_replacer.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    @decorator.cuda_test
+    def test_replace(self):
+        """Test replace function in TeReplacer."""
+        te_transformer = te.TransformerLayer(self.hidden_size, self.ffn_hidden_size, self.num_attention_heads)
+        te_transformer.to(dtype=self.dtype).cuda()
+        model = TeReplacer.replace(te_transformer)
+
+        msamp_modules = []
+
+        def _check_model(model):
```
Comment (wkcn) [inline tests/te/test_replacer.py:38]: Please check whether the weights of linears have been replaced with `ScalingParameter`, and the weight gradients are 
 `ScalingTensor`.

File: msamp/te/replacer.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    @classmethod
+    def _replace(cls, model):
+        for mod in TeReplacer.module_weight_names:
+            if isinstance(model, mod):
+                mod.is_msamp_module = True
+                weight_names = TeReplacer.module_weight_names[mod]
+                for wname in weight_names:
+                    if not hasattr(model, wname):
+                        continue
+                    # assert hasattr(model, wname), [k for k, v in model.named_parameters()]
```
Comment (wkcn) [inline msamp/te/replacer.py:30]: Please remove this comment.

File: pyproject.toml (hunk 1 above)
Comment (tocean) [reply pyproject.toml:41]: I believe TE will not be re-installed if there is already a same version. Just as below:
image

File: msamp/te/replacer.py (hunk 3 above)
Comment (tocean) [reply msamp/te/replacer.py]: Done.

File: tests/te/test_replacer.py (hunk 2 above)
Comment (tocean) [reply tests/te/test_replacer.py:44]: Add logic to check if the weights are `ScalingParameter`. But for weight gradients, looks like their types are still `torch.Tensor` . Can you help confirm this?

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

File: msamp/te/extension.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""MS-AMP te.extension module."""
+
+import torch
+
```
Comment (guoshzhao) [inline msamp/te/extension.py:7]: no blank line.

**Review by guoshzhao — APPROVED** [review]

File: msamp/te/extension.py (hunk 4 above)
Comment (tocean) [reply msamp/te/extension.py]: Fixed.
